rust/lance/src/dataset/write/update.rs

```diff
@@ -0,0 +1,566 @@
+        let expr = planner
+            .parse_filter(filter)
+            .map_err(box_error)
+            .context(InvalidInputSnafu)?;
+        self.condition = Some(
+            planner
+                .optimize_expr(expr)
+                .map_err(box_error)
+                .context(InvalidInputSnafu)?,
+        );
+        Ok(self)
+    }
+
+    pub fn set(mut self, column: impl AsRef<str>, value: &str) -> Result<Self> {
+        let field = self
+            .dataset
+            .schema()
+            .field(column.as_ref())
+            .ok_or_else(|| {
+                Error::invalid_input(
+                    format!(
+                        "Column '{}' does not exist in dataset schema: {:?}",
+                        column.as_ref(),
+                        self.dataset.schema()
+                    ),
+                    location!(),
+                )
+            })?;
+
+        // TODO: support nested column references. This is mostly blocked on the
+        // ability to insert them into the RecordBatch properly.
+        if column.as_ref().contains('.') {
+            return Err(Error::NotSupported {
+                source: format!(
+                    "Nested column references are not yet supported. Referenced: {}",
+                    column.as_ref(),
+                )
+                .into(),
+                location: location!(),
+            });
+        }
+
+        let schema: Arc<ArrowSchema> = Arc::new(self.dataset.schema().into());
+        let planner = Planner::new(schema.clone());
+        let mut expr = planner
+            .parse_expr(value)
+            .map_err(box_error)
+            .context(InvalidInputSnafu)?;
+
+        // Cast expression to the column's data type if necessary.
+        let dest_type = field.data_type();
+        let df_schema = DFSchema::try_from(schema.as_ref().clone())?;
+        let src_type = expr
+            .get_type(&df_schema)
+            .map_err(box_error)
+            .context(InvalidInputSnafu)?;
+        if dest_type != src_type {
+            // TODO: remove this once DataFusion supports casting List to FSL
+            expr = match expr {
+                Expr::Literal(value @ ScalarValue::List(_, _))
+                    if matches!(dest_type, DataType::FixedSizeList(_, _)) =>
+                {
+                    Expr::Literal(safe_coerce_scalar(&value, &dest_type).ok_or_else(|| {
+                        ArrowError::CastError(format!(
+                            "Failed to cast {} to {} during planning",
+                            value.data_type(),
+                            dest_type
+                        ))
+                    })?)
+                }
+                _ => expr
+                    .cast_to(&dest_type, &df_schema)
+                    .map_err(box_error)
+                    .context(InvalidInputSnafu)?,
+            };
+        }
+
+        // Optimize the expression. For example, this might apply the cast on
```

Comment:
Why would it cast the literals if we've already done `safe_coerce_scalar`?

Reply:
You are right that above if we are using `safe_coerce_scalar` above, that's handled. But `expr.cast_to()` (in the other branch) just wraps the expression in a cast, it doesn't actually propagate casts to literals.